cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/resource/volume/AwsVolumeResourceBuilder.java

```diff
@@ -367,21 +371,128 @@ public CloudResource delete(AwsContext context, AuthenticatedContext auth, Cloud
+                    .collect(Collectors.toList());
+            List<Volume> allVolumes = rebuildVolumes(auth, volumeSetAttributes, deletedVolumes, deletedVolumeInfo.getTemplateVolume());
+            volumeSetAttributes.getVolumes().stream()
+                    .filter(volume -> !deletedVolumeInfo.getDeletedVolumeIds().contains(volume.getId()))
+                    .forEach(allVolumes::add);
+            return allVolumes;
+        } catch (Ec2Exception e) {
+            throw e;
+        } catch (ExecutionException | InterruptedException e) {
+            throw new CloudbreakRuntimeException("Cannot recover volumes.", e);
+        }
+    }
+
+    private DeletedVolumeInfo getDeletedVolumeInfo(AmazonEc2Client client, Set<String> allVolumesInVolumeSet, CloudResource resource) {
+        Set<String> deletedVolumeIds = new HashSet<>();
+        software.amazon.awssdk.services.ec2.model.Volume templateVolume = null;
+        for (String volumeId : allVolumesInVolumeSet) {
+            DescribeVolumesRequest describeVolumesRequest = DescribeVolumesRequest.builder()
+                    .volumeIds(volumeId)
+                    .build();
+            LOGGER.debug("Going to describe volume with id: [{}]", String.join(",", describeVolumesRequest.volumeIds()));
+            boolean deleted = false;
+            try {
+                DescribeVolumesResponse result = client.describeVolumes(describeVolumesRequest);
+                ResourceStatus volumeResourceStatus = getResourceStatus(result);
+                deleted = ResourceStatus.DELETED == volumeResourceStatus;
+                if (templateVolume == null && EnumSet.of(ResourceStatus.ATTACHED, ResourceStatus.CREATED).contains(volumeResourceStatus)) {
+                    templateVolume = result.volumes().stream().findFirst().orElse(null);
+                }
+            } catch (Ec2Exception e) {
+                if (!"InvalidVolume.NotFound".equals(e.awsErrorDetails().errorCode())) {
+                    throw e;
+                }
+                LOGGER.info("The volume doesn't need to be deleted as it does not exist on the provider side. Reason: {}", e.getMessage());
+                deleted = true;
+            }
+            if (deleted) {
+                deletedVolumeIds.add(volumeId);
+            }
+        }
+        if (templateVolume == null) {
+            throw new AwsResourceException(String.format("All of the volumes from the set attached to the instance (%s)" +
+                    " are already deleted, cannot recover. " +
+                    "Please try to repair the node with the 'Delete volumes' option.", resource.getInstanceId()),
+                    ResourceType.AWS_VOLUMESET, resource.getName());
+        }
+        return new DeletedVolumeInfo(templateVolume, deletedVolumeIds);
+    }
+
+    private List<Volume> rebuildVolumes(AuthenticatedContext auth, VolumeSetAttributes volumeSetAttributes,
```

Comment:
Can we notify the customer that some of the disks were missing and have been recreated?